File: frontend/lib/api/ratios.ts
/**
 * Ratio engine client (Phase 2C/2D). Read endpoints + admin config CRUD.
 */
import { api } from './client';

export type RatioFormat = 'ratio' | 'percent' | 'dollar' | 'days';

export interface RatioRow {
  key: string;
  label: string;
  category: string;
  format: RatioFormat;
  value: number | null;
  py_value: number | null;
  enabled: boolean;
  threshold_min: number | null;
  threshold_max: number | null;
  threshold_direction: string | null;
  breached: boolean;
  custom?: boolean;
  error?: string | null;
}

export interface RatiosResponse {
  entity_code: string;
  period_label: string;
  period_start: string;
  period_end: string;
  ttm_start: string;
  ttm_end: string;
  context: {
    ttm_ebitda: number;
    ttm_ebitda_excl_dgip: number;
    total_debt: number;
    overdraft_reclassified: number;
    equity_reclassified: number;
    balances_balanced: boolean;
    annual_debt_service: number;
    annual_debt_service_source: string;
    debt_service_breakdown: Record<string, number | string>;
    fixed_charges: number;
    fixed_charges_source: string;
    fixed_charges_breakdown: Record<string, number | string>;
  };
  ratios: RatioRow[];
}

export async function getRatios(entityCode: string, periodEnd?: string): Promise<RatiosResponse> {
  const params: Record<string, string> = { entity_code: entityCode };
  if (periodEnd) params.period_end = periodEnd;
  const res = await api.get<RatiosResponse>('/api/dashboard/ratios', { params });
  return res.data;
}

export interface RoleRow {
  role: string;
  account_code: string;
  account_name: string;
  account_type: string | null;
  account_subtype: string | null;
}

export async function getRatioRoles(entityCode: string): Promise<{ roles: RoleRow[] }> {
  const res = await api.get('/api/dashboard/ratios/roles', { params: { entity_code: entityCode } });
  return res.data;
}

export async function getRatioTokens(entityCode: string): Promise<{
  tokens: { totals: string[]; accounts: string[] };
  builtin_formulas: Record<string, { numerator_expr: string; denominator_expr: string; output_type: string }>;
}> {
  const res = await api.get('/api/dashboard/ratios/tokens', { params: { entity_code: entityCode } });
  return res.data;
}

export interface CustomRatio {
  key: string;
  label: string;
  numerator_expr: string;
  denominator_expr: string | null;
  output_type: string;
  enabled: boolean;
  threshold_min: number | null;
  threshold_max: number | null;
  threshold_direction: string | null;
}

export async function listCustomRatios(entityCode: string): Promise<{ custom: CustomRatio[] }> {
  const res = await api.get('/api/dashboard/ratios/custom', { params: { entity_code: entityCode } });
  return res.data;
}

export async function upsertRatioConfig(entityCode: string, body: {
  ratio_key: string; enabled: boolean;
  threshold_min: number | null; threshold_max: number | null; threshold_direction: string | null;
}) {
  return (await api.put('/api/dashboard/ratios/config', body, { params: { entity_code: entityCode } })).data;
}

export async function setRatioInput(entityCode: string, key: string, value: number) {
  return (await api.put('/api/dashboard/ratios/inputs', { key, value }, { params: { entity_code: entityCode } })).data;
}

export async function clearRatioInput(entityCode: string, key: string) {
  return (await api.delete(`/api/dashboard/ratios/inputs/${key}`, { params: { entity_code: entityCode } })).data;
}

export async function addRatioRole(entityCode: string, role: string, account_code: string) {
  return (await api.post('/api/dashboard/ratios/roles', { role, account_code }, { params: { entity_code: entityCode } })).data;
}

export async function removeRatioRole(entityCode: string, role: string, account_code: string) {
  return (await api.delete('/api/dashboard/ratios/roles', { params: { entity_code: entityCode }, data: { role, account_code } })).data;
}

export async function upsertCustomRatio(entityCode: string, body: CustomRatio) {
  return (await api.post('/api/dashboard/ratios/custom', body, { params: { entity_code: entityCode } })).data;
}

export async function deleteCustomRatio(entityCode: string, key: string) {
  return (await api.delete(`/api/dashboard/ratios/custom/${key}`, { params: { entity_code: entityCode } })).data;
}

export async function previewCustomRatio(entityCode: string, body: {
  key: string; label: string; numerator_expr: string; denominator_expr: string | null; output_type: string;
  enabled: boolean; threshold_min: number | null; threshold_max: number | null; threshold_direction: string | null;
}): Promise<{ ok: boolean; value?: number | null; error?: string }> {
  return (await api.post('/api/dashboard/ratios/custom/preview', body, { params: { entity_code: entityCode } })).data;
}

File: backend/app/routes/ratios.py
"""
Ratio engine endpoints (Phase 2C). Read-only analytics; computes the built-in
ratio library current vs prior-year, with per-entity enable/threshold config.
"""
from __future__ import annotations

from datetime import date as DateType
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import text

from ..db import db_session
from ..services import get_entity_by_code
from ..services_auth import require_role
from ..services_ratios import (
    RATIO_META,
    build_financials_context,
    compute_builtin_ratios,
    get_account_roles,
    account_roles_detail,
    seed_account_roles,
    resolve_annual_debt_service,
)

router = APIRouter(prefix="/api/dashboard/ratios", tags=["ratios"])


def _shift_year(d: DateType, years: int) -> DateType:
    try:
        return d.replace(year=d.year + years)
    except ValueError:
        return d.replace(year=d.year + years, day=28)


def _resolve_period(session, entity_id: str, period_end: str | None) -> dict[str, Any]:
    if period_end:
        row = session.execute(
            text(
                "SELECT period_start, period_end, period_label, status FROM accounting_periods "
                "WHERE entity_id=:e AND period_end=:pe"
            ),
            {"e": entity_id, "pe": period_end},
        ).mappings().first()
        if not row:
            raise HTTPException(404, f"No accounting period ending {period_end}")
        return dict(row)
    # default: latest closed period (the most recent finalized month)
    row = session.execute(
        text(
            "SELECT period_start, period_end, period_label, status FROM accounting_periods "
            "WHERE entity_id=:e AND status='closed_locked' AND period_end<=CURRENT_DATE "
            "ORDER BY period_end DESC LIMIT 1"
        ),
        {"e": entity_id},
    ).mappings().first()
    if not row:
        row = session.execute(
            text(
                "SELECT period_start, period_end, period_label, status FROM accounting_periods "
                "WHERE entity_id=:e AND period_end<=CURRENT_DATE ORDER BY period_end DESC LIMIT 1"
            ),
            {"e": entity_id},
        ).mappings().first()
    if not row:
        raise HTTPException(404, "No accounting periods for this entity")
    return dict(row)


def _breach(value: float | None, cfg: dict[str, Any] | None) -> bool:
    if value is None or not cfg:
        return False
    direction = cfg.get("threshold_direction")
    tmin, tmax = cfg.get("threshold_min"), cfg.get("threshold_max")
    if direction == "min" and tmin is not None and value < float(tmin):
        return True
    if direction == "max" and tmax is not None and value > float(tmax):
        return True
    return False


@router.get("")
def get_ratios(
    entity_code: str = Query(...),
    period_end: str | None = Query(default=None),
    _user: Any = Depends(require_role("viewer")),
) -> dict[str, Any]:
    with db_session() as session:
        entity = get_entity_by_code(session, entity_code)
        if not entity:
            raise HTTPException(404, f"Unknown entity code: {entity_code}")
        eid = entity["id"]
        period = _resolve_period(session, eid, period_end)
        ps, pe = period["period_start"], period["period_end"]
        roles = get_account_roles(session, eid)

        inputs_rows = session.execute(
            text("SELECT key, value FROM entity_ratio_inputs WHERE entity_id=:e"),
            {"e": eid},
        ).mappings().all()
        inputs = {r["key"]: float(r["value"]) for r in inputs_rows}

        # Debt service is GL-derived per TTM window (override from
        # entity_ratio_inputs wins). Resolve separately for current vs PY.
        ads_cur, ads_src, ads_breakdown = resolve_annual_debt_service(session, eid, pe)
        ads_py, _, _ = resolve_annual_debt_service(session, eid, _shift_year(pe, -1))

        cfg_rows = session.execute(
            text(
                "SELECT ratio_key, enabled, threshold_min, threshold_max, threshold_direction "
                "FROM entity_ratio_config WHERE entity_id=:e"
            ),
            {"e": eid},
        ).mappings().all()
        cfg = {r["ratio_key"]: dict(r) for r in cfg_rows}

        ctx = build_financials_context(session, entity_id=eid, period_start=ps, period_end=pe, roles=roles)
        cur = compute_builtin_ratios(ctx, {**inputs, "annual_debt_service": ads_cur})
        ctx_py = build_financials_context(
            session, entity_id=eid, period_start=_shift_year(ps, -1), period_end=_shift_year(pe, -1), roles=roles,
        )
        prior = compute_builtin_ratios(ctx_py, {**inputs, "annual_debt_service": ads_py})

    ratios = []
    for key, meta in RATIO_META.items():
        c = cfg.get(key)
        enabled = c["enabled"] if c else True
        value = cur.get(key)
        ratios.append({
            "key": key,
            "label": meta["label"],
            "category": meta["category"],
            "format": meta["format"],
            "value": value,
            "py_value": prior.get(key),
            "enabled": enabled,
            "threshold_min": float(c["threshold_min"]) if c and c["threshold_min"] is not None else None,
            "threshold_max": float(c["threshold_max"]) if c and c["threshold_max"] is not None else None,
            "threshold_direction": c["threshold_direction"] if c else None,
            "breached": _breach(value, c),
        })

    # cross-period retail ratios (need both windows)
    def growth(a, b):
        return None if not b else (a - b) / b * 100.0
    ratios.append({
        "key": "sales_growth_yoy_pct", "label": "Sales growth (YoY)", "category": "Retail",
        "format": "percent", "value": growth(ctx["revenue"], ctx_py["revenue"]),
        "py_value": None, "enabled": True, "threshold_min": None, "threshold_max": None,
        "threshold_direction": None, "breached": False,
    })
    ratios.append({
        "key": "bank_balance_vs_py_pct", "label": "Bank balance vs prior year", "category": "Retail",
        "format": "percent", "value": growth(ctx["cash"], ctx_py["cash"]),
        "py_value": None, "enabled": True, "threshold_min": None, "threshold_max": None,
        "threshold_direction": None, "breached": False,
    })

    return {
        "entity_code": entity_code,
        "period_label": period["period_label"],
        "period_start": ps.isoformat(),
        "period_end": pe.isoformat(),
        "ttm_start": ctx["ttm_start"],
        "ttm_end": ctx["ttm_end"],
        "context": {
            "ttm_ebitda": ctx["ttm_ebitda"],
            "ttm_ebitda_excl_dgip": ctx["ttm_ebitda_excl_dgip"],
            "total_debt": ctx["total_debt"],
            "overdraft_reclassified": ctx["overdraft_reclassified"],
            "equity_reclassified": ctx["equity_reclassified"],
            "balances_balanced": ctx["balances_balanced"],
            "annual_debt_service": ads_cur,
            "annual_debt_service_source": ads_src,
            "debt_service_breakdown": ads_breakdown,
        },
        "ratios": ratios,
    }


@router.get("/roles")
def list_roles(
    entity_code: str = Query(...),
    _user: Any = Depends(require_role("viewer")),
) -> dict[str, Any]:
    with db_session() as session:
        entity = get_entity_by_code(session, entity_code)
        if not entity:
            raise HTTPException(404, f"Unknown entity code: {entity_code}")
        return {"entity_code": entity_code, "roles": account_roles_detail(session, entity["id"])}


@router.post("/seed-roles")
def seed_roles(
    entity_code: str = Query(...),
    _user: Any = Depends(require_role("admin")),
) -> dict[str, Any]:
    """Initial auto-seed of account roles from QBO type/subtype heuristics.
    Idempotent (ON CONFLICT DO NOTHING); preserves admin edits."""
    with db_session() as session:
        entity = get_entity_by_code(session, entity_code)
        if not entity:
            raise HTTPException(404, f"Unknown entity code: {entity_code}")
        inserted = seed_account_roles(session, entity["id"])
    return {"entity_code": entity_code, "seeded": inserted}
